solidity/contracts/isms/native/OptimismIsm.sol

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+import {IInterchainSecurityModule} from "../../interfaces/IInterchainSecurityModule.sol";
+import {IOptimismMessageHook} from "../../interfaces/hooks/IOptimismMessageHook.sol";
+import {Message} from "../../libs/Message.sol";
+import {TypeCasts} from "../../libs/TypeCasts.sol";
+
+import {ICrossDomainMessenger} from "@eth-optimism/contracts/libraries/bridge/ICrossDomainMessenger.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {CrossChainEnabledOptimism} from "@openzeppelin/contracts/crosschain/optimism/CrossChainEnabledOptimism.sol";
+
+/**
+ * @title OptimismISM
+ * @notice Uses the native Optimism bridge to verify interchain messages.
+ */
+contract OptimismISM is
+    IInterchainSecurityModule,
+    CrossChainEnabledOptimism,
+    Ownable
+{
+    // ============ Constants ============
+
+    // solhint-disable-next-line const-name-snakecase
+    uint8 public constant moduleType =
+        uint8(IInterchainSecurityModule.Types.OPTIMISM);
+
+    // Optimism's native CrossDomainMessenger deployed on L2
+    // @dev Only allowed to call `receiveFromHook`.
+    ICrossDomainMessenger public immutable l2Messenger;
+
+    // ============ Public Storage ============
+
+    // Hook deployed on L1 responsible for sending message via the Optimism bridge
+    IOptimismMessageHook public l1Hook;
+    // mapping to check if the specific messageID from a specific emitter has been received
+    // @dev anyone can send an untrusted messageId, so need to check for that while verifying
+    mapping(bytes32 => mapping(address => bool)) public receivedEmitters;
```

Comment:
should we have an abstract L2 ISM that implements the shared receivedEmitters, hook, and verify logic?